flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java

```diff
@@ -0,0 +1,197 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.strategy;
+
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.scheduler.DeploymentOption;
+import org.apache.flink.runtime.scheduler.ExecutionVertexDeploymentOption;
+import org.apache.flink.runtime.scheduler.SchedulerOperations;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@link SchedulingStrategy} instance for batch job which schedule vertices when input data are ready.
+ */
+public class LazyFromSourcesSchedulingStrategy implements SchedulingStrategy {
+
+	private final SchedulerOperations schedulerOperations;
+
+	private final SchedulingTopology schedulingTopology;
+
+	private final Map<ExecutionVertexID, DeploymentOption> deploymentOptions;
+
+	private final InputDependencyConstraintChecker inputConstraintChecker;
+
+	public LazyFromSourcesSchedulingStrategy(
+			SchedulerOperations schedulerOperations,
+			SchedulingTopology schedulingTopology) {
+		this.schedulerOperations = checkNotNull(schedulerOperations);
+		this.schedulingTopology = checkNotNull(schedulingTopology);
+		this.deploymentOptions = new HashMap<>();
+		this.inputConstraintChecker = new InputDependencyConstraintChecker();
+	}
+
+	@Override
+	public void startScheduling() {
+		List<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions = new ArrayList<>();
+		final DeploymentOption updateOption = new DeploymentOption(true);
+		final DeploymentOption nonUpdateOption = new DeploymentOption(false);
+
+		for (SchedulingExecutionVertex schedulingVertex : schedulingTopology.getVertices()) {
+			DeploymentOption option = nonUpdateOption;
+			for (SchedulingResultPartition srp : schedulingVertex.getProducedResultPartitions()) {
+				if (srp.getPartitionType().isBlocking()) {
+					inputConstraintChecker.addSchedulingResultPartition(srp);
+				} else {
+					option = updateOption;
+				}
+			}
+			deploymentOptions.put(schedulingVertex.getId(), option);
+
+			if (schedulingVertex.getConsumedResultPartitions().isEmpty()) {
+				// schedule vertices without consumed result partition
+				executionVertexDeploymentOptions.add(
+						new ExecutionVertexDeploymentOption(schedulingVertex.getId(), option));
+			}
+		}
+
+		schedulerOperations.allocateSlotsAndDeploy(executionVertexDeploymentOptions);
```

Comment:
There is some common structure in `startScheduling` and `restartTasks`:

1. Initialization of data structures
1. scheduling of _"candidate"_ vertices

At least the 2nd part should be virtually the same:

```
@Override
	public void startScheduling() {
		// ... init, i.e., building the deployment options map, etc.

		allocateSlotsAndDeployExecutionVertexIds(getAllVerticesFromTopology());
	}

	@Override
	public void restartTasks(Set<ExecutionVertexID> verticesToRestart) {
		// ... init, i.e., resetting partitions

		allocateSlotsAndDeployExecutionVertexIds(verticesToRestart);
	}
```

There are also some similarities between `onExecutionStateChange` and `onPartitionConsumable`. Both schedule the consumers that are:

- in created state and
- have a satisfied input dependency 

So the outline of these functions looks something like this:
```
	@Override
	public void onExecutionStateChange(ExecutionVertexID executionVertexId, ExecutionState executionState) {
		if (!ExecutionState.FINISHED.equals(executionState)) {
			return;
		}

		final SchedulingExecutionVertex producer = getSchedulingVertex(executionVertexId);

		List<SchedulingResultPartition> finishedPartitions = new ArrayList<>();
		// ... populate finishedPartitions 

		final Set<SchedulingExecutionVertex> verticesToSchedule = finishedPartitions.stream()
			.flatMap(schedulingResultPartition -> schedulingResultPartition.getConsumers().stream())
			.collect(Collectors.toSet());

		allocateSlotsAndDeployExecutionVertices(verticesToSchedule);
	}

	@Override
	public void onPartitionConsumable(ExecutionVertexID executionVertexId, ResultPartitionID resultPartitionId) {
		final SchedulingResultPartition resultPartition = // ... look up result partition

		if (!resultPartition.getPartitionType().isPipelined()) {
			return;
		}

		final SchedulingExecutionVertex producerVertex = getSchedulingVertex(executionVertexId);

		// ... checkState

		allocateSlotsAndDeployExecutionVertices(resultPartition.getConsumers());
	}
```

We can remove redundancy by moving logic into some helper functions:
```
private SchedulingExecutionVertex getSchedulingVertex(final ExecutionVertexID executionVertexId) {
		return schedulingTopology.getVertex(executionVertexId)
			.orElseThrow(() -> new IllegalStateException("can not find scheduling vertex for " + executionVertexId));
	}

	private void allocateSlotsAndDeployExecutionVertexIds(final Collection<ExecutionVertexID> executionVertexIds) {
		allocateSlotsAndDeployExecutionVertices(executionVertexIds
			.stream()
			.map(this::getSchedulingVertex)
			.collect(Collectors.toList()));
	}

	private void allocateSlotsAndDeployExecutionVertices(final Collection<SchedulingExecutionVertex> schedulingExecutionVertices) {
		schedulerOperations.allocateSlotsAndDeploy(schedulingExecutionVertices
			.stream()
			.filter(isInputConstraintSatisfied().and(isInCreatedExecutionState()))
			.map(SchedulingExecutionVertex::getId)
			.map(executionVertexID -> new ExecutionVertexDeploymentOption(executionVertexID, deploymentOptions.get(executionVertexID)))
			.collect(Collectors.toSet()));
	}

	private Predicate<SchedulingExecutionVertex> isInCreatedExecutionState() {
		return schedulingExecutionVertex -> schedulingExecutionVertex.getState() == ExecutionState.CREATED;
	}

	private Predicate<SchedulingExecutionVertex> isInputConstraintSatisfied() {
		return inputConstraintChecker::check;
	}

	private Set<ExecutionVertexID> getAllVerticesFromTopology() {
		return StreamSupport
			.stream(schedulingTopology.getVertices().spliterator(), false)
			.map(SchedulingExecutionVertex::getId)
			.collect(Collectors.toSet());
	}
```

Reply:
OK